src/Serialize.h

```diff
@@ -33,6 +33,99 @@ namespace drafter {
 
     refract::Registry& GetNamedTypesRegistry();
 
+    template<typename T>
+    struct SectionInfo {
+        typedef SectionInfo<T> Type;
+        typedef T SectionType;
+        typedef snowcrash::SourceMap<SectionType> SourceMapType;
+
+        const SectionType& section;
+        const SourceMapType& sourceMap;
+        const bool empty;
+
+        SectionInfo(const SectionType& section, const SourceMapType& sourceMap) : section(section), sourceMap(sourceMap), empty(false) {}
+        SectionInfo() : section(Type::NullSection()), sourceMap(Type::NullSourceMap()), empty(true) {}
+
+        /**
+         * BE CAREFUL while assign SectionInfo it probably will not work as you expected
+         * but we need this to allow store SectionInfo in containers
+         *
+         * alternative solution is store `section` and `sourceMap` in C++11 smart pointers
+         */
+        SectionInfo<T>& operator=(const SectionInfo<T>& other) { 
+            return *this; 
+        }
+
+        static const SectionType& NullSection() {
+            static SectionType nullSection;
+            return nullSection;
+        }
+
+        static const SourceMapType& NullSourceMap() {
+            static SourceMapType nullSourceMap;
+            return nullSourceMap;
+        }
+
+        bool isNull() const { 
+            return empty; 
+        }
+
+        bool hasSourceMap() const {
+            const SourceMapType& null = NullSourceMap();
+            return &sourceMap != &null;
+        }
+    };
+
+    template <typename T>
+    SectionInfo<T> MakeSectionInfo(const T& section, const snowcrash::SourceMap<T>& sourceMap, const bool hasSourceMap)
+    {
+        return SectionInfo<T>(section, hasSourceMap ? sourceMap : SectionInfo<T>::NullSourceMap());
+    }
+
+
+    template <typename T>
+    SectionInfo<T> MakeSectionInfoWithoutSourceMap(const T& section)
+    {
+        return SectionInfo<T>(section, SectionInfo<T>::NullSourceMap());
+    }
+
+#define MAKE_SECTION_INFO(from, member) MakeSectionInfo(from.section.member, from.sourceMap.member, from.hasSourceMap())
+
+    template<typename ResultType, typename Collection1, typename Collection2, typename BinOp>
+    ResultType Zip(const Collection1& collection1, const Collection2& collection2, const BinOp& Combinator) {
+        ResultType result;
+        std::transform(collection1.begin(), collection1.end(), collection2.begin(), std::back_inserter(result), Combinator);
+        return result;
+    }
+
+    template<typename T>
+    struct SectionInfoCollection  : public std::vector<SectionInfo<typename T::value_type> > {
```

Comment:
Sure. I start in `Serialize.h` while it was small class, whule develop it grow up